### PR TITLE
Support REPLACE INTO for INSERT statements

### LINKED ISSUE #12515: Support REPLACE INTO for INSERT statements
### Is your feature request related to a problem or challenge?

It would be great to support `INSERT OR REPLACE INTO` statements as it is a common SQL query.

### Describe the solution you'd like

A new argument to `TableProvider::insert_into` so that `TableProvider` implementors can implement `REPLACE INTO` when needed.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by fmeringdal**
## Which issue does this PR close?

Closes #12515 .

## Are these changes tested?

Yes, in `datafusion/core/tests/user_defined/insert_operation.rs`.

## Are there any user-facing changes?

Yes, `overwrite` boolean flags has been replaced by `InsertOp` enum in the following APIs and configs: `TableProvider::insert_into`, `FileSinkConfig`, `DataFrameWriteOptions`.

### CODE REVIEW

File: datafusion/catalog/src/table.rs [hunk 1]
```diff
@@ -273,6 +273,7 @@ pub trait TableProvider: Sync + Send {
         _state: &dyn Session,
         _input: Arc<dyn ExecutionPlan>,
         _overwrite: bool,
+        _replace_into: bool,
```
Comment (alamb) [inline datafusion/catalog/src/table.rs:276]: Can you please document this new parameter and explain what it means (bonus points for documenting what `_overwrite` means too)

**Review by alamb — COMMENTED** [review]
Thank you for this contirbution @fmeringdal  -- this code looks good to me. I am worried about the lack of tests, however

Basically if there isn't test coverage we could potentially break this feature without realizing it in some future refactor. 

maybe we could add something in 

https://github.com/apache/datafusion/tree/main/datafusion/core/tests/user_defined

That shows that a mock up user defined `TableProvider` will have `insert_into` called with the correct value of `replace_into` when the relevant SQL is planned 🤔 

It is unfortunate we don't seem to already have such a test

**Comment by fmeringdal** [comment]
Thanks for the quick review @alamb 🙏 I will get some tests added.

I wanted to run an idea by you before I implement the change to restructure the `WriteOp` enum from:

```rs
pub enum WriteOp {
    InsertOverwrite,
    InsertReplace,
    InsertInto,
    Delete,
    Update,
    Ctas,
}
```

To:
```rs
pub enum WriteOp {
    Insert(InsertOp),
    Delete,
    Update,
    Ctas,
}

pub enum InsertOp {
    Append,    // Represents a regular INSERT INTO operation
    Overwrite, // Represents an INSERT OVERWRITE operation
    Replace,   // Represents an INSERT OR REPLACE operation
}
```

The purpose of such a change would be so that `TableProvider::insert_into` could take an `op: InsertOp` argument instead of two boolean arguments for `overwrite` and `replace_into` . I think it makes it slightly more convenient for implementors of the `insert_into` method to match on an enum rather than checking two booleans.
What do you think?

**Comment by alamb** [comment]
> The purpose of such a change would be so that TableProvider::insert_into could take an op: InsertOp argument instead of two boolean arguments for overwrite and replace_into . I think it makes it slightly more convenient for implementors of the insert_into method to match on an enum rather than checking two booleans.
What do you think?

I agree it would be a good change 👌 

Thank you!

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

File: datafusion/catalog/src/table.rs (hunk 1 above)
Comment (fmeringdal) [reply datafusion/catalog/src/table.rs]: > That shows that a mock up user defined TableProvider will have insert_into called with the correct value of replace_into when the relevant SQL is planned

Test added in `datafusion/core/tests/user_defined/insert_operation.rs`.

> Can you please document this new parameter and explain what it means (bonus points for documenting what _overwrite means too)

Documentation added to the variants of the `InsertOp` enum.

**Comment by fmeringdal** [comment]
Ready for another round of review! Changes applied since the first iteration:
- `InsertOp` enum has been introduced and replaced the `overwrite` bool. This introduced a few breaking changes as described in the PR description. Any other places I need to document these breaking changes?
- Test added to `datafusion/core/tests/user_defined/insert_operation.rs`.
- Handle both `REPLACE INTO ...` and `INSERT OR REPLACE INTO ...`

File: datafusion/core/tests/user_defined/insert_operation.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    error::Result,
+    prelude::{SessionConfig, SessionContext},
+};
+use datafusion_catalog::{Session, TableProvider};
+use datafusion_expr::{dml::InsertOp, Expr, TableType};
+use datafusion_physical_expr::{EquivalenceProperties, Partitioning};
+use datafusion_physical_plan::{DisplayAs, ExecutionMode, ExecutionPlan, PlanProperties};
+
+#[tokio::test]
+async fn insert_operation_is_passed_correctly_to_table_provider() {
```
Comment (alamb) [inline datafusion/core/tests/user_defined/insert_operation.rs:32]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @fmeringdal  for this PR and for making it easy to review the PR (nicely documented, nicely written code, and nicely tested) 🏆 

I had this PR checked out anyways so I took the liberty of merging up from main to resolve the conflicts, as well as added a license notice to the file and fixed a clippy issue.

**Comment by alamb** [comment]
Merged up to get CI fix / hopefully pass

**Comment by alamb** [comment]
🚀

**Pull request merged**